src/Http/Routing/src/Matching/DataSourceDependentMatcher.cs

```diff
@@ -16,12 +15,17 @@ internal class DataSourceDependentMatcher : Matcher
 
         public DataSourceDependentMatcher(
             EndpointDataSource dataSource,
+            Lifetime lifetime,
             Func<MatcherBuilder> matcherBuilderFactory)
         {
             _matcherBuilderFactory = matcherBuilderFactory;
 
             _cache = new DataSourceDependentCache<Matcher>(dataSource, CreateMatcher);
             _cache.EnsureInitialized();
+
+            // This will Dispose the cache when the lifetime is disposed, this allows
+            // the service provider to manage the lifetime of the cache.
+            lifetime.Cache = _cache;
```

Comment:
lifetime being null will result in NRE. That may be by design, just making sure it has been thought of.